kpi/tests/utils/dicts.py

```diff
@@ -0,0 +1,35 @@
+from __future__ import annotations
+
+
+def nested_dict_from_keys(dict_: dict) -> dict:
+    """
+    Transforms a dictionary with keys containing slashes into a nested
+    dictionary structure.
+    """
+
+    result = {}
+
+    for key, value in dict_.items():
+        keys = key.split('/')
+        sub_dict = result
+        for sub_key in keys[:-1]:
+            if sub_key not in sub_dict:
+                sub_dict[sub_key] = {}
+            sub_dict = sub_dict[sub_key]
+
+        if isinstance(value, list):
+            sub_dict[keys[-1]] = [
+                {
+                    sub_key.split('/')[-1]: sub_val
+                    for sub_key, sub_val in item.items()
+                }
+                for item in value if item
+            ]
+        else:
+            sub_dict[keys[-1]] = (
+                nested_dict_from_keys(value)
+                if isinstance(value, dict)
+                else value
+            )
+
+    return result
```

Comment:
This utility should probably have its own unit test, maybe going all the way through to creating XML from the nested dict.

As written currently, I'm not sure it does what you want. I made this example with nested repeating groups; here's the submission JSON:
```json
{
  "_id": 48872,
  "formhub/uuid": "8a368610e429400ca4a197ffaa832998",
  "start": "2024-09-20T00:13:30.387-04:00",
  "end": "2024-09-20T00:16:17.295-04:00",
  "outer_q": "outer 1",
  "middle_group": [
    {
      "middle_group/middle_q": "middle 1.1",
      "middle_group/inner_group": [
        {
          "middle_group/inner_group/inner_q": "inner 1.1.1"
        },
        {
          "middle_group/inner_group/inner_q": "inner 1.1.2"
        }
      ]
    },
    {
      "middle_group/middle_q": "middle 1.2",
      "middle_group/inner_group": [
        {
          "middle_group/inner_group/inner_q": "inner 1.2.1"
        },
        {
          "middle_group/inner_group/inner_q": "inner 1.2.2"
        }
      ]
    }
  ],
  "__version__": "vaRBXEHb7ReCxvxKFBGHfa",
  "meta/instanceID": "uuid:9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d",
  "_xform_id_string": "ab8ySkQP4RcdmrzeFzLReL",
  "_uuid": "9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d",
  "_attachments": [],
  "_status": "submitted_via_web",
  "_geolocation": [
    null,
    null
  ],
  "_submission_time": "2024-09-20T04:16:50",
  "_tags": [],
  "_notes": [],
  "_validation_status": {},
  "_submitted_by": null
}
```
When I run it through `nested_dict_from_keys()`, I get:
```python
{
    '__version__': 'vaRBXEHb7ReCxvxKFBGHfa',
    '_attachments': [],
    '_geolocation': [],
    '_id': 48872,
    '_notes': [],
    '_status': 'submitted_via_web',
    '_submission_time': '2024-09-20T04:16:50',
    '_submitted_by': None,
    '_tags': [],
    '_uuid': '9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d',
    '_validation_status': {},
    '_xform_id_string': 'ab8ySkQP4RcdmrzeFzLReL',
    'end': '2024-09-20T00:16:17.295-04:00',
    'formhub': {'uuid': '8a368610e429400ca4a197ffaa832998'},
    'meta': {'instanceID': 'uuid:9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d'},
    'middle_group': [
        {
            'inner_group': [
                {'middle_group/inner_group/inner_q': 'inner ' '1.1.1'},
                {'middle_group/inner_group/inner_q': 'inner ' '1.1.2'},
            ],
            'middle_q': 'middle 1.1',
        },
        {
            'inner_group': [
                {'middle_group/inner_group/inner_q': 'inner ' '1.2.1'},
                {'middle_group/inner_group/inner_q': 'inner ' '1.2.2'},
            ],
            'middle_q': 'middle 1.2',
        },
    ],
    'outer_q': 'outer 1',
    'start': '2024-09-20T00:13:30.387-04:00',
}
```

Reply:
I wish I had written this comment better :sweat_smile: Let's make a decently complex example submission and get its JSON and XML representations from the API. We can then add a unit test for `nested_dict_from_keys()` itself that runs the JSON through both `nested_dict_from_keys()` and pyxform's `dict2xform()`, and finally verifies the resulting XML against what was expected.

The current problem is that for
```json
 "middle_group": [
    {
      "middle_group/inner_group": [
        {
          "middle_group/inner_group/inner_q": "some response"
        },
```
…we get
```xml
<middle_group>
  <inner_group>
    <middle_group_inner_group_inner_q>some response
```
…but we need
```xml
<middle_group>
  <inner_group>
    <inner_q>some response
```

We can make a failing unit test to catch this and hopefully other funky nuances in the translation of submission JSON to XML.


## Full XML from the examples in my previous comment

When I pass the JSON (above) produced by `nested_dict_from_keys()` into `dict2xform(…, 'wow_amazing_id_string')` from pyxform, I get:
```xml
<?xml version="1.0" encoding="utf-8"?>
   <wow_amazing_id_string id="wow_amazing_id_string">
<__version__>vaRBXEHb7ReCxvxKFBGHfa</__version__>
<_attachments></_attachments>
<_geolocation></_geolocation>
<_id>48872</_id>
<_notes></_notes>
<_status>submitted_via_web</_status>
<_submission_time>2024-09-20T04:16:50</_submission_time>
<_submitted_by>None</_submitted_by>
<_tags></_tags>
<_uuid>9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d</_uuid>
<_validation_status></_validation_status>
<_xform_id_string>ab8ySkQP4RcdmrzeFzLReL</_xform_id_string>
<end>2024-09-20T00:16:17.295-04:00</end>
<formhub>
  <uuid>8a368610e429400ca4a197ffaa832998</uuid>
</formhub>
<meta>
  <instanceID>uuid:9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d</instanceID>
</meta>
<middle_group>
  <inner_group>
    <middle_group_inner_group_inner_q>inner 1.1.1</middle_group_inner_group_inner_q>
  </inner_group>
  <inner_group>
    <middle_group_inner_group_inner_q>inner 1.1.2</middle_group_inner_group_inner_q>
  </inner_group>
  <middle_q>middle 1.1</middle_q>
</middle_group>
<middle_group>
  <inner_group>
    <middle_group_inner_group_inner_q>inner 1.2.1</middle_group_inner_group_inner_q>
  </inner_group>
  <inner_group>
    <middle_group_inner_group_inner_q>inner 1.2.2</middle_group_inner_group_inner_q>
  </inner_group>
  <middle_q>middle 1.2</middle_q>
</middle_group>
<outer_q>outer 1</outer_q>
<start>2024-09-20T00:13:30.387-04:00</start>
</wow_amazing_id_string>
```

What we want (focusing on the repeating group) is:
```xml
<ab8ySkQP4RcdmrzeFzLReL id="ab8ySkQP4RcdmrzeFzLReL" version="1 (2024-09-20 04:12:43)">
  <formhub>
    <uuid>8a368610e429400ca4a197ffaa832998</uuid>
  </formhub>
  <start>2024-09-20T00:13:30.387-04:00</start>
  <end>2024-09-20T00:16:17.295-04:00</end>
  <outer_q>outer 1</outer_q>
  <middle_group>
    <middle_q>middle 1.1</middle_q>
    <inner_group>
      <inner_q>inner 1.1.1</inner_q>
    </inner_group>
    <inner_group>
      <inner_q>inner 1.1.2</inner_q>
    </inner_group>
  </middle_group>
  <middle_group>
    <middle_q>middle 1.2</middle_q>
    <inner_group>
      <inner_q>inner 1.2.1</inner_q>
    </inner_group>
    <inner_group>
      <inner_q>inner 1.2.2</inner_q>
    </inner_group>
  </middle_group>
  <__version__>vaRBXEHb7ReCxvxKFBGHfa</__version__>
  <meta>
    <instanceID>uuid:9dc9f510-d3c7-4c9a-a35a-91d6d8fd1f3d</instanceID>
  </meta>
</ab8ySkQP4RcdmrzeFzLReL>
```

Reply:
Tests added...